src/best_offset_prefetcher.cpp

```diff
@@ -0,0 +1,269 @@
+    // return all offsets to 0
+    reset_offsets();
+    // set current round back to 0
+    this->current_round = 0;
+    // set the test offset back to the end index (for timely prefetches)
+    this->test_offset_index = 0;
+}
+
+uint64_t BestOffsetPrefetcher::find_max_score(){
+    std::pair<uint64_t,uint64_t>  max_score = std::pair<uint64_t,uint64_t>(1, BAD_SCORE); 
+    for (auto s : offset_scores){
+        if (s.second >= max_score.second){
+            // set the offset to the offset value
+            max_score.first = s.first;
+            // set the second to the score
+            max_score.second = s.second; 
+        } 
+    }
+    return max_score.first; 
+}
+
+void BestOffsetPrefetcher::move_test_offset_ptr(){
+    // if the current offset is the total minus 1
+    if (this->test_offset_index == (NUM_OFFSETS - 1) ){
+        // reset the pointer and set the round up
+        this->test_offset_index = 0;
+        this->current_round++;
+    // if max out the nuber of rounds, 
+        // reset the prefetcher
+         if (this->current_round == ( RND_MAX + 1 ) ){
+            this->current_offset = find_max_score(); 
+            reset_prefetcher();
+        }
+    }
+     else{
+        // move address in test_offset_point along
+        this->test_offset_index++;
+    }
+}
+
+
+void BestOffsetPrefetcher::learn(uint64_t addr, uint64_t cycle){
+    // offset to currently test
+    uint64_t test_offset = (this->offset_scores[this->test_offset_index]).first;
+    // if we're going off the page, dont' get it
+    uint64_t base_addr = addr - test_offset;
+    // if we're going off the page, dont' get it
+    if ( ((addr - test_offset) < 0 )
+        or
+        (
+        // if trying to check on another page, don't do it
+        (addr & PAGE_MASK) != ((base_addr) & PAGE_MASK)
+        )){
+        // move this offset along
+        move_test_offset_ptr();
+        return;
+    }
+
+    // check if address in the recent requests table
+    if (this->recent_requests.exists(base_addr, cycle) == true){
+        // if increasing that index by 1 would put us over the max score
+            // reset the prefetcher
+        if (this->offset_scores[this->test_offset_index].second == (MAX_SCORE-1)){
+            print_scores();
+            // increase the score for that offset
+            this->current_offset = this->offset_scores[this->test_offset_index].first;
+            reset_prefetcher();
+            return;
+        }
+        // otherwise, if found the address but not maxed out, increment
+            // its score
+        else{
+            this->offset_scores[this->test_offset_index].second++; 
+        } 
+    }
+    move_test_offset_ptr();
+}
+
+BestOffsetPrefetcher::BestOffsetPrefetcher(const g_string& _name,
+                                       const g_string& _target,
+                                       bool _monitor_GETS, bool _monitor_GETX,
+                                       uint32_t _degree):
```

Comment:
degree is not used and in fact for BOP I don't think it makes sense. Can you add something like:
if (degree != 1)
std::cout << "Ignoring degree parameter, using default degree of 1" <<std::endl;